### PR TITLE
feat: add evm_snapshot/revert

### LINKED ISSUE #69: evm_api: Implement `evm_snapshot` API endpoint
### **Description:**
The `evm_snapshot` endpoint needs to be implemented. The purpose of this endpoint is to snapshot the state of the blockchain at the current block and return the ID (to be used for reverting state later). [Ganache docs for reference](https://ganache.dev/#evm_snapshot).

```rust
// Interface in the `trait`
#[rpc(name = "evm_snapshot")]
fn snapshot(
    &self,
) -> BoxFuture<Result<U64>>;

// Actual code in the `impl`
fn snapshot(
    &self,
) -> BoxFuture<Result<U64>> {
    // TODO: Solution here
}
```

#### Expected response
Command returns proper response:
```
curl -X POST http://localhost:8011 \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "evm_snapshot",
       "params": []
     }'
```

Response:
```
{"jsonrpc":"2.0", "result":"0x1", "id":1}
```

### **Requirements:**

1. **Implement the API:**
    - The endpoint should return a populated `U64` of the snapshot's ID.
    - Ensure the method executes the call reliably and efficiently.
    - Ensure errors are properly returned.

2. **Unit Tests:**
    - Write unit tests to verify the functionality of the `snapshot` method.
    - Ensure coverage for multiple edge-case scenarios.

3. **Documentation:**
    - Add inline Rust documentation (`///` comments) detailing the method's purpose, return values, and any potential edge cases or error scenarios.
    - Update `SUPPORTED_APIS.md` with updated documentation and examples.
    - (Optional) Add request to `test_endpoints.http` for easier manual verification.

### **Task Checklist:**
- [ ] Implement the `snapshot` method.
- [ ] Write unit tests for the `snapshot` method.
- [ ] Add inline documentation for the `evm_snapshot` endpoint.
- [ ] Update documentation for the `evm_snapshot` endpoint in `SUPPORTED_APIS.md`.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* Adds support for `evm_snapshot` and `evm_revert`

# Why :hand:
* Useful for developers to snapshot/rever the evm state while debugging or performing tests.

# Evidence :camera:
image
image

# Notes :memo:
* Fixes #69 && #70

### CODE REVIEW

File: src/evm.rs [hunk 1]
```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
...
+                into_jsrpc_error(Web3Error::InternalError)
+            })?;
+            let snapshot_id_index = snapshot_id.as_usize().saturating_sub(1);
+            if snapshot_id_index >= snapshots.len() {
+                log::error!("no snapshot exists for the id '{}'", snapshot_id);
+                return Err(into_jsrpc_error(Web3Error::InternalError));
+            }
+
+            // remove all snapshots following the index and use the first snapshot for restore
+            let selected_snapshot = snapshots
```
Comment (MexicanAce) [inline src/evm.rs:234]: Just to confirm my understanding, we are keeping the snapshots as a FILO stack. So here's what that would look like:

1. Create snapshot A, we return ID `1`
2. Create snapshot B, we return ID `2`
3. Revert to snapshot A -> Snapshot B is deleted

File: src/evm.rs [hunk 2]
```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
...
+        let inner = Arc::clone(&self.node);
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            let mut snapshots = snapshots.write().map_err(|err| {
+                log::error!("failed acquiring read lock for snapshots: {:?}", err);
+                into_jsrpc_error(Web3Error::InternalError)
+            })?;
+            let snapshot_id_index = snapshot_id.as_usize().saturating_sub(1);
+            if snapshot_id_index >= snapshots.len() {
```
Comment (MexicanAce) [inline src/evm.rs:228]: nit/suggestion: We can also check for negative values here to avoid potential errors down below

File: src/evm.rs (hunk 1 above)
Comment (nbaztec) [reply src/evm.rs:234]: Yes, this is the same behavior as in [ganache](https://ganache.dev/#evm_revert)

File: src/evm.rs (hunk 2 above)
Comment (nbaztec) [reply src/evm.rs:228]: `saturating_sub` should clip the operation at `0`. It's pretty much impossible to get to the negative range or underflow, since we're dealing with unsigned arithmetic and `U64` request deserializers.

File: src/evm.rs [hunk 3]
```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
...
+            // remove all snapshots following the index and use the first snapshot for restore
+            let selected_snapshot = snapshots
+                .drain(snapshot_id_index..)
+                .next()
+                .expect("unexpected failure, value must exist");
+
+            inner
+                .write()
+                .map_err(|err| format!("failed acquiring read lock for snapshots: {:?}", err))
+                .and_then(|mut writer| writer.restore_snapshot(selected_snapshot).map(|_| true))
```
Comment (MexicanAce) [inline src/evm.rs:242]: suggestion: We should add a log statement when the node is reverting to a snapshot

```suggestion
                .and_then(|mut writer| {
                    log::info!("Reverting node to snapshot {}", snapshot_id);
                    writer.restore_snapshot(selected_snapshot).map(|_| true)
                })
```

File: src/evm.rs [hunk 4]
```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
...
+                })
+                .and_then(|reader| reader.snapshot())
+                .map_err(|err| {
+                    log::error!("failed creating snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })?;
+            snapshots
+                .write()
+                .map(|mut snapshots| {
+                    snapshots.push(snapshot);
```
Comment (MexicanAce) [inline src/evm.rs:207]: suggestion: If we add a log ling notifying the developer that the node has reverted to a snapshot ID, then I think we should also add a log statement when the snapshot is taken
```suggestion
                    log::info!("Reverting node to snapshot {}", snapshots.len() - 1);
                    snapshots.push(snapshot);
```

**Review by MexicanAce — APPROVED** [review]
LGTM, left 2 nit comments for additional logging

**Review by MexicanAce — APPROVED** [review]
LGTM
